File: advent-of-code-2023/day8/main.py
import sys

sys.setrecursionlimit(10000000)

def solution():
    with open('input.txt', 'r') as f:
        lines = f.readlines()
        instruction = lines[0].strip()
        lines = lines[2:]
        graph = {}
        for line in lines:
            start, end = [i.strip() for i in line.strip().split('=')]
            left, right = [i.strip() for i in end.strip().split(',')]
            left = left[1:]
            right = right[:-1]
            graph[start] = (left, right)

        def solve(node, instruction, step):
            if node == "ZZZ":
                return 1
            
            left, right = graph[node]
            if step == len(instruction):
                step = 0
            
            if instruction[step] == "L":
                return 1+solve(left, instruction, step+1)
            else:
                return 1+solve(right, instruction, step+1)
        
        return solve("AAA", instruction, 0) - 1


def solution2():
    with open('input.txt', 'r') as f:
        lines = f.readlines()
        instruction = lines[0].strip()
        lines = lines[2:]
        graph = {}
        for line in lines:
            start, end = [i.strip() for i in line.strip().split('=')]
            left, right = [i.strip() for i in end.strip().split(',')]
            left = left[1:]
            right = right[:-1]
            graph[start] = (left, right)

        def solve2(nodes, instruction, step):
            end_with_z = [i[-1]=="Z" for i in nodes]
            if all(end_with_z):
                return 1
            
            node_left, node_right = [[graph[node][0] for node in nodes], [graph[node][1] for node in nodes]]
            if step == len(instruction):
                step = 0
            
            if instruction[step] == "L":
                return 1+solve2(node_left, instruction, step+1)
            else:
                return 1+solve2(node_right, instruction, step+1)
        
        start_nodes = [node for node in graph if node[-1]=="A"]
        return solve2(start_nodes, instruction, 0) - 1



if __name__ == '__main__':
    print(solution2())

File: advent-of-code-2023/day8/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	fmt.Println(solution2())
}

func solve(nodes []string, instruction string, step int, graph map[string][]string, i int) int {
	if i == 1_000_000 {
		fmt.Println(nodes, step, i)
		return 0
	}
	flag := true
	for _, node := range nodes {
		if node[len(node)-1] != 'Z' {
			flag = false
			break
		}
	}
	if flag {
		return 1
	}

	node_left := make([]string, 0)
	node_right := make([]string, 0)
	for _, node := range nodes {
		node_left = append(node_left, graph[node][0])
		node_right = append(node_right, graph[node][1])
	}
	if step == len(instruction) {
		step = 0
	}

	if instruction[step] == 'L' {
		return solve(node_left, instruction, step+1, graph, i+1) + 1
	}

	return solve(node_right, instruction, step+1, graph, i+1) + 1
}

func solution2() int {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lines := make([]string, 0)

	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	instruction := strings.TrimSpace(lines[0])
	lines = lines[2:]
	graph := make(map[string][]string)
	for _, line := range lines {
		line = strings.TrimSpace(line)
		start_end := strings.Split(line, "=")
		start := strings.TrimSpace(start_end[0])
		end := strings.TrimSpace(start_end[1])
		left := strings.Split(end, ",")[0]
		right := strings.Split(end, ",")[1]
		left = strings.TrimSpace(left[1:])
		right = strings.TrimSpace(right[:len(right)-1])
		graph[start] = []string{left, right}
	}
	start_nodes := make([]string, 0)
	for k := range graph {
		if k[len(k)-1] == 'A' {
			start_nodes = append(start_nodes, k)
		}
	}

	// return solve(start_nodes, instruction, 0, graph, i) - 1

	var index int
	var i int
	for {
		if i%1_000_000_0 == 0 {
			fmt.Println(start_nodes, i, index)
		}
		for idx, node := range start_nodes {
			if instruction[index%len(instruction)] == 'L' {
				start_nodes[idx] = graph[node][0]
			} else {
				start_nodes[idx] = graph[node][1]
			}
		}
		index++
		i++

		flag := true
		for _, node := range start_nodes {
			if node[len(node)-1] != 'Z' {
				flag = false
			}
		}

		if flag {
			break
		}
	}

	return i

}
